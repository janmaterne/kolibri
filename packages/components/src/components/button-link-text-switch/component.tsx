/* eslint-disable jsx-a11y/label-has-associated-control */
import { Component, h, Host, JSX, Prop } from '@stencil/core';

import { ButtonOrLinkOrTextWithChildrenProps, ButtonWithChildrenProps, LinkWithChildrenProps, TextWithChildrenProps } from '../../types/button-link-text';
import { Props } from './types';
import { KolButtonWcTag, KolLinkWcTag, KolSpanWcTag } from '../../core/component-names';

/**
 * Internal component that renders an action or text component like a button or a link.
 * @internal
 */
@Component({
	tag: 'kol-button-link-text-switch-wc',
	shadow: false,
})
export class KolButtonLinkTextSwitchWc implements Props {
	public render(): JSX.Element {
		return <Host class="kol-button-link-text-switch-wc">{this.renderContent()}</Host>;
	}

	private renderContent() {
		if ((this._link as ButtonWithChildrenProps)._on) {
			const button = this._link as ButtonWithChildrenProps;
			return this.button(button);
		} else if ((this._link as LinkWithChildrenProps)._href) {
			const link = this._link as LinkWithChildrenProps;
			return this.link(link);
		} else {
			const text = this._link as TextWithChildrenProps;
			return this.text(text);
		}
	}

	private button = (button: ButtonWithChildrenProps): JSX.Element => <KolButtonWcTag {...button}></KolButtonWcTag>;

	private link = (link: LinkWithChildrenProps): JSX.Element => <KolLinkWcTag {...link}></KolLinkWcTag>;

	private text = (text: TextWithChildrenProps): JSX.Element => <KolSpanWcTag {...text}></KolSpanWcTag>;

	/**
	 * Die Link-Daten welche diese Komponente verwendet, um die entsprechende Komponente zu rendern.
	 */
	@Prop() public _link!: ButtonOrLinkOrTextWithChildrenProps;
}
