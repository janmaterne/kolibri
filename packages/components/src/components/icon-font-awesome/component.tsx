import type { Generic } from 'adopted-style-sheets';
import { Component, h, Host, JSX, Prop } from '@stencil/core';

import { FontAwesome, FontAwesomeOssPrefix } from '../../enums/font-awesome';
import { PropAriaLabel } from '../../types/props/aria-label';
import { KolIconTag } from '../../core/component-names';

type RequiredProps = PropAriaLabel & {
	icon: FontAwesome;
	prefix: FontAwesomeOssPrefix;
};
type OptionalProps = {
	part: string;
};
// type Props = Generic.Element.Members<RequiredProps, OptionalProps>;

/**
 * @deprecated Wir empfehlen die flexiblere KolIcon-Komponente zu verwenden.
 */
@Component({
	tag: 'kol-icon-font-awesome',
	shadow: false,
})
export class KolIconFontAwesome implements Generic.Element.Members<RequiredProps, OptionalProps> {
	public render(): JSX.Element {
		return (
			<Host class="kol-icon-font-awesome">
				<KolIconTag
					exportparts={`icon${typeof this._part === 'string' ? `,${this._part}` : ''}`}
					_ariaLabel={this._ariaLabel}
					_icon={typeof this._prefix === 'string' && typeof this._icon === 'string' ? `${this._prefix} fa-${this._icon}` : (undefined as unknown as string)}
				/>
			</Host>
		);
	}

	/**
	 * Setzt die semantische Beschriftung der Komponente.
	 */
	@Prop() public _ariaLabel!: string;

	/**
	 * Gibt einen Identifier eines Font Awesome Icons an. (https://fontawesome.com/v5.15/icons)
	 */
	@Prop() public _icon!: FontAwesome;

	/**
	 * Gibt den Prefix der Font Awesome Icons an. (https://fontawesome.com/v5.15/how-to-use/on-the-web/referencing-icons/basic-use)
	 */
	@Prop() public _prefix!: FontAwesomeOssPrefix;

	/**
	 * Deprecated: Gibt den Identifier für den CSS-Part an, um das Icon von Außen ändern zu können. (https://meowni.ca/posts/part-theme-explainer/)
	 *
	 * @deprecated Das Styling sollte stets über CSS erfolgen.
	 */
	@Prop() public _part?: string;
}
