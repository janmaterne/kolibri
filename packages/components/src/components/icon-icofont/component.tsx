import type { Generic } from 'adopted-style-sheets';
import { Component, h, Host, JSX, Prop } from '@stencil/core';

import { Icofont } from '../../types/icofont';
import { PropAriaLabel } from '../../types/props/aria-label';
import { KolIconTag } from '../../core/component-names';

type RequiredProps = PropAriaLabel & {
	icon: Icofont;
};
type OptionalProps = {
	part: string;
};
// type Props = Generic.Element.Members<RequiredProps, OptionalProps>;

/**
 * @deprecated Wir empfehlen die flexiblere KolIcon-Komponente zu verwenden.
 */
@Component({
	tag: 'kol-icon-icofont',
	shadow: false,
})
export class KolIconIcofont implements Generic.Element.Members<RequiredProps, OptionalProps> {
	public render(): JSX.Element {
		return (
			<Host class="kol-icon-icofont">
				<KolIconTag
					exportparts={`icon${typeof this._part === 'string' ? `,${this._part}` : ''}`}
					_ariaLabel={this._ariaLabel}
					_icon={typeof this._icon === 'string' ? `icofont-${this._icon}` : (undefined as unknown as string)}
				/>
			</Host>
		);
	}

	/**
	 * Setzt die semantische Beschriftung der Komponente.
	 */
	@Prop() public _ariaLabel!: string;

	/**
	 * Gibt einen Identifier eines Icofont Icons an. (https://icofont.com/icons)
	 */
	@Prop() public _icon!: Icofont;

	/**
	 * Deprecated: Gibt den Identifier für den CSS-Part an, um das Icon von Außen ändern zu können. (https://meowni.ca/posts/part-theme-explainer/)
	 *
	 * @deprecated Das Styling sollte stets über CSS erfolgen.
	 */
	@Prop() public _part?: string;
}
