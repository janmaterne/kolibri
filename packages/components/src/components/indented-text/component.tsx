import type { IndentedTextAPI, IndentedTextStates } from '../../schema';
import type { JSX } from '@stencil/core';
import { Component, h, Host, State } from '@stencil/core';

/**
 * @slot - Der Text.
 */
@Component({
	tag: 'kol-indented-text-wc',
	shadow: false,
})
export class KolIndentedTextWc implements IndentedTextAPI {
	@State() public state: IndentedTextStates = {};

	public render(): JSX.Element {
		return (
			<Host class="kol-indented-text-wc">
				<div>
					{/* TODO: Für was wird dieses div benötigt? */}
					<slot />
				</div>
			</Host>
		);
	}
}
