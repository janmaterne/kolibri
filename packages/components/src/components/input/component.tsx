/* eslint-disable jsx-a11y/label-has-associated-control */
import { Component, Element, Fragment, h, Host, JSX, Prop } from '@stencil/core';

import { translate } from '../../i18n';
import { Stringified } from '../../types/common';
import { AnyIconFontClass, KoliBriCustomIcon, KoliBriHorizontalIcons } from '../../types/icons';
import { IdPropType } from '../../types/props/id';
import { LabelWithExpertSlotPropType } from '../../types/props/label';
import { SuggestionsPropType } from '../../types/props/suggestions';
import { TooltipAlignPropType } from '../../types/props/tooltip-align';
import { W3CInputValue } from '../../types/w3c';
import { handleSlotContent, showExpertSlot } from '../../utils/reuse';
import { FormFieldMsg } from '../@shared/form-field-msg';
import { Props as ButtonProps } from '../button/types';
import { Props } from './types';
import { KolButtonWcTag, KolIconTag, KolTooltipWcTag } from '../../core/component-names';

/**
 * @internal
 */
@Component({
	tag: 'kol-input-wc',
	shadow: false,
})
export class KolInputWc implements Props {
	@Element() private readonly host?: HTMLElement;

	private slotName: string = 'input';

	public componentWillRender(): void {
		this.slotName = this._slotName ? this._slotName : 'input';
	}

	private catchInputSlot = (slot?: HTMLDivElement): void => {
		handleSlotContent(this.host!, slot!, this.slotName);
	};

	private getIconsProp(): KoliBriHorizontalIcons | undefined {
		return this._icons || this._icon;
	}

	private getIconStyles(icon?: AnyIconFontClass | KoliBriCustomIcon): Record<string, string> {
		return icon && typeof icon === 'object' && icon.style ? icon.style : {};
	}

	public render(): JSX.Element {
		const hasError = !this._readOnly && typeof this._error === 'string' && this._error.length > 0 && this._touched === true;
		const hasExpertSlot = showExpertSlot(this._label);
		const hasHint = typeof this._hint === 'string' && this._hint.length > 0;
		const useTooltopInsteadOfLabel = !hasExpertSlot && this._hideLabel;

		return (
			<Host
				class={{
					'kol-input-wc': true,
					disabled: this._disabled === true,
					error: hasError === true,
					'read-only': this._readOnly === true,
					required: this._required === true,
					touched: this._touched === true,
					'hidden-error': this._hideError === true,
				}}
			>
				<label class="input-label" id={!useTooltopInsteadOfLabel ? `${this._id}-label` : undefined} hidden={useTooltopInsteadOfLabel} htmlFor={this._id}>
					{/* INFO: span is needed for css styling :after content like a star (*) or optional text ! */}
					<span>
						{/* INFO: label comes with any html tag or as plain text! */}
						<slot name="label"></slot>
					</span>
				</label>
				{hasHint && (
					<span class="hint" id={`${this._id}-hint`}>
						{this._hint}
					</span>
				)}
				<div
					class={{
						input: true,
						'icon-left': typeof this.getIconsProp()?.left === 'object',
						'icon-right': typeof this.getIconsProp()?.right === 'object',
					}}
				>
					{this.getIconsProp()?.left && (
						<KolIconTag
							_ariaLabel=""
							_icons={(this.getIconsProp()?.left as KoliBriCustomIcon).icon}
							style={this.getIconStyles(this.getIconsProp()?.left)}
						></KolIconTag>
					)}
					<div ref={this.catchInputSlot} id={this.slotName} class="input-slot"></div>
					{typeof this._smartButton === 'object' && this._smartButton !== null && (
						<KolButtonWcTag
							_customClass={this._smartButton._customClass}
							_disabled={this._smartButton._disabled}
							_icons={this._smartButton._icons}
							_hideLabel={true}
							_id={this._smartButton._id}
							_label={this._smartButton._label}
							_on={this._smartButton._on}
							_tooltipAlign={this._smartButton._tooltipAlign}
							_variant={this._smartButton._variant}
						></KolButtonWcTag>
					)}
					{this.getIconsProp()?.right && (
						<KolIconTag
							_ariaLabel=""
							_icons={(this.getIconsProp()?.right as KoliBriCustomIcon).icon}
							style={this.getIconStyles(this.getIconsProp()?.right)}
						></KolIconTag>
					)}
				</div>
				{useTooltopInsteadOfLabel && (
					<KolTooltipWcTag
						/**
						 * Dieses Aria-Hidden verhindert das doppelte Vorlesen des Labels,
						 * verhindert aber nicht das Aria-Labelledby vorgelesen wird.
						 */
						aria-hidden="true"
						class="input-tooltip"
						_align={this._tooltipAlign}
						_id={this._hideLabel ? `${this._id}-label` : undefined}
						_label={this._label}
					></KolTooltipWcTag>
				)}
				{hasError && <FormFieldMsg _alert={this._alert} _hideError={this._hideError} _error={this._error} _id={this._id} />}
				{Array.isArray(this._suggestions) && this._suggestions.length > 0 && (
					<datalist id={`${this._id}-list`}>
						{this._suggestions.map((option: W3CInputValue) => (
							<option value={option} />
						))}
					</datalist>
				)}
				{this._hasCounter && (
					<span class="counter" aria-atomic="true" aria-live="polite">
						{this._currentLength}
						{this._maxLength && (
							<Fragment>
								<span aria-label={translate('kol-of')} role="img">
									/
								</span>
								{this._maxLength}
							</Fragment>
						)}{' '}
						<span>{translate('kol-characters')}</span>
					</span>
				)}
			</Host>
		);
	}

	/**
	 * Defines whether the screen-readers should read out the notification.
	 * @TODO: Change type back to `AlertPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _alert?: boolean = true;

	/**
	 * @internal
	 */
	@Prop() public _currentLength?: number;

	/**
	 * Makes the element not focusable and ignore all events.
	 * @TODO: Change type back to `DisabledPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _disabled?: boolean = false;

	/**
	 * Defines the error message text.
	 */
	@Prop() public _error?: string = '';

	/**
	 * Shows the character count on the lower border of the input.
	 * @TODO: Change type back to `HasCounterPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hasCounter?: boolean = false;

	/**
	 * Hides the error message but leaves it in the DOM for the input's aria-describedby.
	 * @TODO: Change type back to `HideErrorPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideError?: boolean = false;

	/**
	 * Hides the caption by default and displays the caption text with a tooltip when the
	 * interactive element is focused or the mouse is over it.
	 * @TODO: Change type back to `HideLabelPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideLabel?: boolean = false;

	/**
	 * Defines the hint text.
	 */
	@Prop() public _hint?: string = '';

	/**
	 * @deprecated Use _icons.
	 */
	@Prop() public _icon?: KoliBriHorizontalIcons;

	/**
	 * Defines the icon classnames (e.g. `_icons="fa-solid fa-user"`).
	 */
	@Prop() public _icons?: KoliBriHorizontalIcons;

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id!: IdPropType;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.). Set to `false` to enable the expert slot.
	 */
	@Prop() public _label!: LabelWithExpertSlotPropType;

	/**
	 * Defines the maximum number of input characters.
	 */
	@Prop() public _maxLength?: number;

	/**
	 * Makes the input element read only.
	 * @TODO: Change type back to `ReadOnlyPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _readOnly?: boolean = false;

	/**
	 * Gibt an, ob die Komponente kein Label rendern soll.
	 */
	@Prop() public _renderNoLabel?: boolean = false;

	/**
	 * Makes the input element required.
	 * @TODO: Change type back to `RequiredPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _required?: boolean = false;

	/**
	 * Ermöglicht den Slotnamen zu bestimmen. Wird nur verwendet, wenn sonst mehrere Slots mit dem gleichen Namen innerhalb eines Shadow DOMs existieren würden.
	 * @internal
	 */
	@Prop() public _slotName?: string;

	/**
	 * Suggestions to provide for the input.
	 */
	@Prop() public _suggestions?: SuggestionsPropType;

	/**
	 * Allows to add a button with an arbitrary action within the element (_hide-label only).
	 */
	@Prop() public _smartButton?: Stringified<ButtonProps>;

	/**
	 * Defines where to show the Tooltip preferably: top, right, bottom or left.
	 */
	@Prop() public _tooltipAlign?: TooltipAlignPropType = 'top';

	/**
	 * Shows if the input was touched by a user.
	 * @TODO: Change type back to `TouchedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _touched?: boolean = false;
}
