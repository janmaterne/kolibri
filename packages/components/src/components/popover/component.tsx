import { arrow, computePosition, flip, MiddlewareData, offset, Placement, shift } from '@floating-ui/dom';
import { Component, h, Host, JSX, Prop, State, Watch } from '@stencil/core';

import { AlignPropType, validateAlign } from '../../types/props/align';
import { ShowPropType, validateShow } from '../../types/props/show';
import { getDocument } from '../../utils/dev.utils';
import { processEnv } from '../../utils/reuse';
import { API, States } from './types';

/**
 * @slot - Der Inhalt des Popover.
 */
@Component({
	tag: 'kol-popover-wc',
	styleUrl: './style.scss',
	shadow: false,
})
export class KolPopoverWc implements API {
	private arrowElement?: HTMLDivElement;
	private popoverElement?: HTMLDivElement;
	private triggerElement?: HTMLElement | null;
	private host?: HTMLElement;

	/* Popover functions */
	private alignPopover = (callBack?: () => unknown): void => {
		setTimeout(() => {
			if (processEnv !== 'test' && this.triggerElement && this.popoverElement) {
				const trigger = this.triggerElement;
				const popoverEl = this.popoverElement;
				const arrowEl = this.arrowElement;

				const middleware = [offset(arrowEl?.offsetHeight ?? 10), flip(), shift()];
				if (arrowEl) {
					middleware.push(arrow({ element: arrowEl }));
				}

				void computePosition(trigger, popoverEl, {
					placement: this.state._align,
					middleware: middleware,
				}).then(({ x, y, middlewareData, placement }) => {
					this.setPosition(x, y, middlewareData, placement, callBack);
				});
			}
		});
	};
	private setPosition(x: number, y: number, middlewareData: MiddlewareData, placement: Placement, callBack?: () => unknown) {
		if (this.popoverElement) {
			const oldPos = {
				left: this.popoverElement.style.left,
				top: this.popoverElement.style.top,
			};
			Object.assign(this.popoverElement.style, {
				left: `${x}px`,
				top: `${y}px`,
			});

			if (this.arrowElement && middlewareData.arrow) {
				switch (placement) {
					case 'top':
						this.arrowElement.style.inset = `100% auto auto ${middlewareData.arrow.x || 0}px`;
						this.arrowElement.style.translate = '0 -50%';
						break;
					case 'right':
						this.arrowElement.style.inset = `${middlewareData.arrow.y || 0}px 100% auto auto`;
						this.arrowElement.style.translate = '50% 0';
						break;
					case 'bottom':
						this.arrowElement.style.inset = `auto auto 100% ${middlewareData.arrow.x || 0}px`;
						this.arrowElement.style.translate = '0 50%';
						break;
					case 'left':
						this.arrowElement.style.inset = `${middlewareData.arrow.y || 0}px auto auto 100%`;
						this.arrowElement.style.translate = '-50% 0';
						break;
				}
			}
			if (oldPos.left !== this.popoverElement.style.left || oldPos.top !== this.popoverElement.style.top) {
				this.alignPopover(callBack);
			} else if (typeof callBack === 'function') {
				callBack();
			}
		}
	}
	private showPopover = (): void => {
		this.addListenersToBody();

		this.alignPopover(() => {
			this.state = { ...this.state, _visible: true };
		});
	};
	private hidePopover(): void {
		this.state = {
			...this.state,
			_visible: false,
		};
		this._show = false;
		this.triggerElement?.focus();
		this.removeListenersToBody();
	}
	private hidePopoverByEscape = (event: KeyboardEvent): void => {
		if (event.key === 'Escape') this.hidePopover();
	};
	private hidePopoverByClickOutside = (event: MouseEvent): void => {
		if (this.host && !this.host.contains(event.target as HTMLElement)) {
			this.hidePopover();
		}
	};

	/* EventListener functions */
	private addListenersToBody(): void {
		const body = getDocument().body;
		body.addEventListener('keyup', this.hidePopoverByEscape);
		body.addEventListener('click', this.hidePopoverByClickOutside);
		document.scrollingElement?.addEventListener('scroll', this.showPopover, { passive: true });
	}
	private removeListenersToBody(): void {
		const body = getDocument().body;
		body.removeEventListener('keyup', this.hidePopoverByEscape);
		body.removeEventListener('click', this.hidePopoverByClickOutside);
		document.scrollingElement?.removeEventListener('scroll', this.showPopover);
	}

	/* catchElement functions */
	private catchHostAndTriggerElement = (element: HTMLElement | null): void => {
		if (element) {
			this.host = element;
			this.triggerElement = element.previousElementSibling as HTMLElement | null;
		}
	};
	private catchPopoverElement = (element?: HTMLDivElement): void => {
		this.popoverElement = element;
	};
	private catchArrowElement = (element?: HTMLDivElement): void => {
		this.arrowElement = element;
	};

	public render(): JSX.Element {
		return (
			<Host ref={this.catchHostAndTriggerElement} class="kol-popover-wc">
				<div class={{ popover: true, hidden: !this.state._show, show: this.state._visible }} ref={this.catchPopoverElement}>
					<div class={`arrow ${this.state._align}`} ref={this.catchArrowElement} />
					<slot />
				</div>
			</Host>
		);
	}

	/**
	 * Defines the alignment of the tooltip, popover or tabs in relation to the element.
	 */
	@Prop() public _align?: AlignPropType = 'top';

	/**
	 * Makes the element show up.
	 * @TODO: Change type back to `ShowPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _show?: boolean = false;

	@State() public state: States = {
		_align: 'top',
		_show: false,
		_visible: false,
	};

	@Watch('_align')
	public validateAlign(value?: AlignPropType): void {
		validateAlign(this, value);
	}

	@Watch('_show')
	public validateShow(value?: ShowPropType): void {
		validateShow(this, value);
		if (value) this.showPopover();
	}

	public componentWillLoad(): void {
		this.validateAlign(this._align);
		this.validateShow(this._show);
	}
}
