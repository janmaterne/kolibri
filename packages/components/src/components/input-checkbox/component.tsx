import { Component, Element, h, Host, JSX, Prop, State, Watch } from '@stencil/core';

import { Stringified } from '../../types/common';
import { InputTypeOnDefault } from '../../types/input/types';
import { LabelWithExpertSlotPropType } from '../../types/props/label';
import { StencilUnknown } from '../../types/unknown';
import { nonce } from '../../utils/dev.utils';
import { stopPropagation, tryToDispatchKoliBriEvent } from '../../utils/events';
import { propagateFocus } from '../../utils/reuse';
import { getRenderStates } from '../input/controller';
import { InputCheckboxController } from './controller';
import { API, InputCheckboxIcon, InputCheckboxVariant, States } from './types';
import { CheckedPropType } from '../../types/props/checked';
import { IndeterminatePropType } from '../../types/props/indeterminate';
import { SyncValueBySelectorPropType } from '../../types/props/sync-value-by-selector';
import { IdPropType } from '../../types/props/id';
import { NamePropType } from '../../types/props/name';
import { TooltipAlignPropType } from '../../types/props/tooltip-align';

/**
 * @slot - Die Beschriftung der Checkbox.
 */
@Component({
	tag: 'kol-input-checkbox',
	styleUrls: {
		default: './style.css',
	},
	shadow: true,
})
export class KolInputCheckbox implements API {
	@Element() private readonly host?: HTMLKolInputCheckboxElement;
	private ref?: HTMLInputElement;

	private readonly catchRef = (ref?: HTMLInputElement) => {
		this.ref = ref;
		propagateFocus(this.host, this.ref);
	};

	public render(): JSX.Element {
		const { ariaDescribedBy } = getRenderStates(this.state);
		const hasExpertSlot = this.state._label === false; // _label="" or _label

		return (
			<Host>
				<kol-input
					class={{
						checkbox: true,
						[this.state._variant]: true,
						'hide-label': !!this.state._hideLabel,
						checked: this.state._checked,
					}}
					data-role={this.state._variant === 'button' ? 'button' : undefined}
					onKeyPress={this.state._variant === 'button' ? this.onChange : undefined}
					tabIndex={this.state._variant === 'button' ? 0 : undefined}
					_alert={this.state._alert}
					_disabled={this.state._disabled}
					_error={this.state._error}
					_hideLabel={this.state._hideLabel}
					_hint={this.state._hint}
					_id={this.state._id}
					_required={this.state._required}
					_touched={this.state._touched}
				>
					{/*  TODO: der folgende Slot ohne Name muss später entfernt werden */}
					<span slot="label">{hasExpertSlot ? <slot></slot> : this.state._label}</span>
					<div slot="input">
						<kol-icon
							onClick={this.onChange}
							_ariaLabel=""
							_icon={this.state._indeterminate ? this.state._icon.indeterminate : this.state._checked ? this.state._icon.checked : this.state._icon.unchecked}
						/>
						<input
							ref={this.catchRef}
							title=""
							accessKey={this.state._accessKey} // by checkbox?!
							aria-describedby={ariaDescribedBy.length > 0 ? ariaDescribedBy.join(' ') : undefined}
							aria-label={this.state._hideLabel && typeof this.state._label === 'string' ? this.state._label : undefined}
							checked={this.state._checked}
							disabled={this.state._disabled}
							id={this.state._id}
							indeterminate={this.state._indeterminate}
							name={this.state._name}
							required={this.state._required}
							tabIndex={this.state._tabIndex}
							type="checkbox"
							{...this.controller.onFacade}
							onChange={this.onChange}
							onClick={undefined} // onClick is not needed since onChange already triggers the correct event
						/>
						<kol-tooltip-wc
							/**
							 * Dieses Aria-Hidden verhindert das doppelte Vorlesen des Labels,
							 * verhindert aber nicht das Aria-Labelledby vorgelesen wird.
							 */
							aria-hidden="true"
							class="input-tooltip"
							hidden={hasExpertSlot || !this.state._hideLabel}
							_align={this._tooltipAlign}
							_label={typeof this.state._label === 'string' ? this.state._label : ''}
						></kol-tooltip-wc>
					</div>
				</kol-input>
			</Host>
		);
	}

	private readonly controller: InputCheckboxController;

	/**
	 * Defines which key combination can be used to trigger or focus the interactive element of the component.
	 */
	@Prop() public _accessKey?: string;

	/**
	 * Defines whether the screen-readers should read out the notification.
	 */
	@Prop({ mutable: true, reflect: true }) public _alert?: boolean = true;

	/**
	 * Defines whether the checkbox is checked or not. Can be read and written.
	 * @TODO: Change type back to `CheckedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _checked?: boolean = false;

	/**
	 * Makes the element not focusable and ignore all events.
	 * @TODO: Change type back to `DisabledPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _disabled?: boolean;

	/**
	 * Defines the error message text.
	 */
	@Prop() public _error?: string;

	/**
	 * Hides the label and shows the description in a Tooltip instead.
	 * @TODO: Change type back to `HideLabelPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideLabel?: boolean;

	/**
	 * Defines the hint text.
	 */
	@Prop() public _hint?: string = '';

	/**
	 * Defines the icon classnames (e.g. `_icon="fa-solid fa-user"`).
	 */
	@Prop() public _icon?: Stringified<InputCheckboxIcon>;

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id?: IdPropType;

	/**
	 * Puts the checkbox in the indeterminate state, does not change the value of _checked.
	 * @TODO: Change type back to `IndeterminatePropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _indeterminate?: boolean;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.). Set to `false` to enable the expert slot.
	 */
	@Prop() public _label?: LabelWithExpertSlotPropType;

	/**
	 * Defines the technical name of an input field.
	 */
	@Prop() public _name?: NamePropType;

	/**
	 * Gibt die EventCallback-Funktionen für das Input-Event an.
	 */
	@Prop() public _on?: InputTypeOnDefault;

	/**
	 * Makes the input element required.
	 * @TODO: Change type back to `RequiredPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _required?: boolean;

	/**
	 * Selector for synchronizing the value with another input element.
	 * @internal
	 */
	@Prop() public _syncValueBySelector?: SyncValueBySelectorPropType;

	/**
	 * Defines which tab-index the primary element of the component has. (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)
	 */
	@Prop() public _tabIndex?: number;

	/**
	 * Defines where to show the Tooltip preferably: top, right, bottom or left.
	 */
	@Prop() public _tooltipAlign?: TooltipAlignPropType = 'top';

	/**
	 * Shows if the input was touched by a user.
	 * @TODO: Change type back to `TouchedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _touched?: boolean = false;

	/**
	 * Deprecated: Defines which variant should be used for presentation.
	 *
	 * @deprecated Verwende stattdessen das Attribute _variant.
	 */
	@Prop() public _type?: InputCheckboxVariant;

	/**
	 * Defines the value of the input.
	 */
	@Prop() public _value?: Stringified<StencilUnknown> = true;

	/**
	 * Defines which variant should be used for presentation.
	 */
	@Prop() public _variant?: InputCheckboxVariant; // TODO: = 'default'; in v2 setzen

	@State() public state: States = {
		_checked: false,
		_icon: {
			checked: 'codicon codicon-check',
			indeterminate: 'codicon codicon-remove',
			unchecked: 'codicon codicon-add',
		},
		_id: `id-${nonce()}`,
		_indeterminate: false,
		_label: false, // ⚠ required
		_value: true,
		_variant: 'default',
	};

	public constructor() {
		this.controller = new InputCheckboxController(this, 'input-checkbox', this.host);
	}

	@Watch('_accessKey')
	public validateAccessKey(value?: string): void {
		this.controller.validateAccessKey(value);
	}

	@Watch('_alert')
	public validateAlert(value?: boolean): void {
		this.controller.validateAlert(value);
	}

	@Watch('_checked')
	public validateChecked(value?: CheckedPropType): void {
		this.controller.validateChecked(value);
	}

	@Watch('_disabled')
	public validateDisabled(value?: boolean): void {
		this.controller.validateDisabled(value);
	}

	@Watch('_error')
	public validateError(value?: string): void {
		this.controller.validateError(value);
	}

	@Watch('_hideLabel')
	public validateHideLabel(value?: boolean): void {
		this.controller.validateHideLabel(value);
	}

	@Watch('_hint')
	public validateHint(value?: string): void {
		this.controller.validateHint(value);
	}

	@Watch('_icon')
	public validateIcon(value?: Stringified<InputCheckboxIcon>): void {
		this.controller.validateIcon(value);
	}

	@Watch('_id')
	public validateId(value?: string): void {
		this.controller.validateId(value);
	}

	@Watch('_indeterminate')
	public validateIndeterminate(value?: IndeterminatePropType): void {
		this.controller.validateIndeterminate(value);
	}

	@Watch('_label')
	public validateLabel(value?: LabelWithExpertSlotPropType): void {
		this.controller.validateLabel(value);
	}

	@Watch('_name')
	public validateName(value?: string): void {
		this.controller.validateName(value);
	}

	@Watch('_on')
	public validateOn(value?: InputTypeOnDefault): void {
		this.controller.validateOn(value);
	}

	@Watch('_required')
	public validateRequired(value?: boolean): void {
		this.controller.validateRequired(value);
	}

	@Watch('_syncValueBySelector')
	public validateSyncValueBySelector(value?: SyncValueBySelectorPropType): void {
		this.controller.validateSyncValueBySelector(value);
	}

	@Watch('_tabIndex')
	public validateTabIndex(value?: number): void {
		this.controller.validateTabIndex(value);
	}

	@Watch('_touched')
	public validateTouched(value?: boolean): void {
		this.controller.validateTouched(value);
	}

	@Watch('_type')
	public validateType(value?: InputCheckboxVariant): void {
		this.controller.validateType(value);
	}

	@Watch('_value')
	public validateValue(value?: Stringified<StencilUnknown>): void {
		this.controller.validateValue(value);
	}

	@Watch('_variant')
	public validateVariant(value?: InputCheckboxVariant): void {
		this.controller.validateVariant(value);
	}

	public componentWillLoad(): void {
		this._alert = this._alert === true;
		this._touched = this._touched === true;
		this.controller.componentWillLoad();
	}

	private onChange = (event: Event): void => {
		this._checked = !this._checked;
		this._indeterminate = false;

		const value = this._checked ? this.state._value : null;

		// Event handling
		stopPropagation(event);
		tryToDispatchKoliBriEvent('change', this.host, value);

		// Static form handling
		this.controller.setFormAssociatedValue(value);
		this.controller.setFormAssociatedCheckboxValue;

		// Callback
		if (typeof this.state._on?.onChange === 'function') {
			this.state._on.onChange(event, value);
		}
	};
}
