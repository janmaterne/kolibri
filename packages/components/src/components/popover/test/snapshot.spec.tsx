import { KolPopoverWcTag } from '../../../core/component-names';
import type { PopoverProps } from '../../../schema';
import { executeSnapshotTests } from '../../../utils/testing';

import { KolPopover } from '../component';

executeSnapshotTests<PopoverProps>(KolPopoverWcTag, [KolPopover], [...['top', 'right', 'bottom', 'left'].map((_align) => ({ _align }) as PopoverProps)]);
