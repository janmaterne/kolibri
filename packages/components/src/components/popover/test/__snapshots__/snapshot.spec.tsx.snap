// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kol-popover-wc should render with _align="bottom" 1`] = `
<kol-popover-wc class="kol-popover">
  <!---->
  <div class="popover" hidden="">
    <div class="arrow bottom"></div>
  </div>
</kol-popover-wc>
`;

exports[`kol-popover-wc should render with _align="left" 1`] = `
<kol-popover-wc class="kol-popover">
  <!---->
  <div class="popover" hidden="">
    <div class="arrow left"></div>
  </div>
</kol-popover-wc>
`;

exports[`kol-popover-wc should render with _align="right" 1`] = `
<kol-popover-wc class="kol-popover">
  <!---->
  <div class="popover" hidden="">
    <div class="arrow right"></div>
  </div>
</kol-popover-wc>
`;

exports[`kol-popover-wc should render with _align="top" 1`] = `
<kol-popover-wc class="kol-popover">
  <!---->
  <div class="popover" hidden="">
    <div class="arrow top"></div>
  </div>
</kol-popover-wc>
`;
